src/VestedRewardsDistribution.sol

```diff
@@ -162,48 +167,65 @@ contract RewardsDistribution {
+     * @dev The amount calculation is delegated to `calc.getMaxAmount()`.
      *  - If the returned value is `0`, the distribution will fail.
      *  - If the returned value is greater than the current unpaid amount, the distributed amount will the latter.
+     * @return amount The amount being distributed.
      */
-    function distribute() external {
-        require(vestId != INVALID_VEST_ID, "RewardsDistribution/invalid-vest-id");
-        require(dssVest.unpaid(vestId) > 0, "RewardsDistribution/empty-vest");
-
-        uint256 total = dssVest.tot(vestId);
-        uint256 finish = dssVest.fin(vestId);
-        uint256 cliff = dssVest.clf(vestId);
-        // If `lastVestedAt == 0`, it means it this is the first time we call `distribute` for the current `vestId`.
-        uint256 prev = lastVestedAt == 0 ? cliff : lastVestedAt;
-
-        uint256 amount = calc.getAmount(block.timestamp, prev, total, finish, cliff);
-        require(amount > 0, "RewardsDistribution/no-pending-amount");
+    function distribute() external returns (uint256 amount) {
+        require(vestId != INVALID_VEST_ID, "VestedRewardsDistribution/invalid-vest-id");
 
+        uint256 unpaid = dssVest.unpaid(vestId);
+        uint256 maxAmount = _getMaxAmount();
         lastVestedAt = block.timestamp;
-
-        dssVest.vest(vestId, amount);
-        require(gem.transfer(address(stakingRewards), amount), "RewardsDistribution/transfer-failed");
+        dssVest.vest(vestId, maxAmount);
+
+        // `dssVest.vest()` sadly does not return the actual amount of vested tokens.
+        // Also it is not safe to query the gem balance of this contract because it might have dangling tokens.
+        // The easiest way is to replicate the internal logic of DssVest here.
+        amount = _min(unpaid, maxAmount);
+        require(amount > 0, "VestedRewardsDistribution/no-pending-amount");
+        require(gem.transfer(address(stakingRewards), amount), "VestedRewardsDistribution/transfer-failed");
         stakingRewards.notifyRewardAmount(amount);
 
         emit Distribute(amount);
     }
+
+    /**
+     * @notice Gets the max amount to pull from the vesting stream for the next distribution.
+     * @dev If `calc` is set, it delegates the calculation to that contract.
+     *      Otherwise, it returns the default amount `type(uint256.max)`.
+     * @return The max amount of tokens.
+     */
+    function _getMaxAmount() internal view returns (uint256) {
+        if (calc == address(0)) {
+            return type(uint256).max;
+        }
+
+        uint256 tot = dssVest.tot(vestId);
+        uint256 fin = dssVest.fin(vestId);
+        // We ensure that `clf == bgn`, so we could use either one.
+        uint256 clf = dssVest.clf(vestId);
+        // If `lastVestedAt == 0`, it means it this is the first time we call `distribute` for the current `vestId`.
+        uint256 prev = lastVestedAt == 0 ? clf : lastVestedAt;
```

Comment:
Yeah, that was safeguarded before within the `DistributionCalc` logic, but now there might be a problem.

Actually, I think the rule should probably be `block.timestamp > prev`.

Reply:
Errm, actually I think this is not needed.
If you call `distribute()` once, `lastVestedAt` is updated to the current block timestamp.

If you call it again within the same block, it will fail at:
```
require(amount > 0, "VestedRewardsDistribution/no-pending-amount");
```
on L194.

However, we might want to add the check for clarity.

Reply:
FYI I added the check.

Reply:
Fixed.